ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,394 @@
+* check if the record can be applied (i.e. concurrent update guard)
+* apply the business logic  
+* write updates to the state
+* write events and (optionally) follow-up commands to continue the processing
+* apply side-effects (e.g. send a response)
+
+![command-processing](ZEP-wf-stream-processing/event-sourcing-processing.png)
+
+The **events** are processed only on replay and rebuild the state. The replay of an event includes:
+* write updates to the state
+
+![event-replay](ZEP-wf-stream-processing/event-sourcing-replay.png)
+
+In order to rebuild the state with the events, it is necessary that all state changes from the command processing can be applied with the events only. So, the command processing may not update the state without writing an event that reflects the state changes.   
+
+When the replay of the events is done then the processing of the commands starts again. 
+
+## Examples
+
+The following examples show the sequence of records for a certain situation and explain the differences to the current behavior. 
+
+### Create Workflow Instance
+
+Command to Process | Written Events and Follow-up Commands
+---|---
+WorkflowInstanceCreationIntent:Create | WorkflowInstanceCreationIntent:Created, Process:Element_Activating
+Process:Element_Activating | Process:Element_Activated, StartEvent:Element_Activating
+StartEvent:Element_Activating | StartEvent:Element_Activated, StartEvent:Element_Completing 
+StartEvent:Element_Completing | StartEvent:Element_Completed, Sequence_Flow_Taken, ServiceTask:Element_Activating 
+
+* the command `WorkflowInstanceCreationIntent:Create` is the initial command and start of the workflow instance
+* the event `Sequence_Flow_Taken` is produced by the command `StartEvent:Element_Completing` and not by an explicit command for the sequence flow (like `Take_Sequence_Flow`) because it can't be interrupted (i.e. no wait state)
+* since the command `StartEvent:Element_Completing` produces the event `Sequence_Flow_Taken`, it must produce the next command `ServiceTask:Element_Activating` to activate the target activity of the taken sequence flow
+
+### Service Task
+
+Command to Process | Written Events and Follow-up Commands
+---|---
+ServiceTask:Element_Activating | ServiceTask:Element_Activated, Variable:Created, Job:Created
+JobBatch:Activate | JobBatch:Activated 
+Job:Complete | Job:Completed, ServiceTask:Element_Completing
+ServiceTask:Element_Completing | Variable: Created, ServiceTask:Element_Completed, Sequence_Flow_Taken, EndEvent:Element_Activating 
+
+* if the service task has an input variable mapping then it produces one or more `Variable:Created` events for the new local variables
+* the event `Job:Created` is produced by the command `ServiceTask:Element_Activating` instead of a `Job:Create` command because it can't be interrupted
+* if the job is completed with variables then these variables will be stored in the state until they are used for the output mapping (i.e. temporal/transient variables)
```

Comment:
This sounds like an interesting case concerning replay

Reply:
Please explain the replay situation in some more detail. As discussed, Job:Complete and Job:Completed concerning temporary variables.

Reply:
We will do this when we expand the ref spec